## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
# Select the first row of the DataFrame
df.iloc[0]


first                                    Donald
last                                    Harding
job                                Immunologist
phone                        793-904-7725x39308
email          harding.donald.7185@sullivan.com
client_id                                 58515
order_id                                8953482
order_date                           2023-04-28
order_week                                   17
order_year                                 2023
item_id                          EUD29711-63-6U
category                                  decor
subcategory                            wall art
unit_price                               1096.8
unit_cost                                762.71
unit_weight                                 7.5
qty                                         105
line_number                                   1
Name: 0, dtype: object

In [5]:
# Use this space to do any additional research
# What are the unique category names
df["category"].unique()

array(['decor', 'consumables', 'software', 'furniture', 'electronics'],
      dtype=object)

In [6]:
# Use this space to do any additional research
# What are the unique subcategory names
df["subcategory"].unique()

array(['wall art', 'pens', 'project management', 'misc', 'plants',
       'kitchen supplies', 'paper products', 'seating', 'computers',
       'stands', 'tables', 'bathroom supplies', 'audio visual',
       'accounting', 'logistics', 'telecom', 'desks', 'cabling'],
      dtype=object)

In [7]:
# Use this space to do any additional research
# Identify incomplete rows (Comment Post-Run:  There aren't any.)
df.count()

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [8]:
# Use this space to do any additional research
# Review datatypes by column (Post-Run:  OK)
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [9]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [10]:
# For the category with the most entries,
# which subcategory had the most entries?
# Create a new df with the category of interest, then count subcategories
cat_df = df.loc[df['category'] == 'consumables']
df['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [11]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [12]:
# Store the client ids of those top 5 clients in a list.
# Source:  https://stackoverflow.com/questions/52441836/how-to-make-a-list-of-values-based-on-value-count-in-pandas
top_five_clients = df['client_id'].value_counts().index[:5].tolist()
print(top_five_clients)



[33615, 66037, 46820, 38378, 24741]


In [13]:
# How many total units (the qty column) did the
# client with the most entries order order?

# Part 1 :: Set the client number to a variable
top_client_number = df['client_id'].value_counts().index[:1].tolist()
top_client_number = top_client_number[0]
print(top_client_number)

33615


In [14]:
# Part 2 :: Set a new dataframe with topclient only
topclient_df = df.loc[(df["client_id"] == top_client_number)]
topclient_df.head()

first   last                                           job  \
5     Jessica  Reyes  Production designer, theatre/television/film   
12    Jessica  Reyes  Production designer, theatre/television/film   
724   Jessica  Reyes  Production designer, theatre/television/film   
840   Jessica  Reyes  Production designer, theatre/television/film   
1405  Jessica  Reyes  Production designer, theatre/television/film   

           phone                     email  client_id  order_id  order_date  \
5     5016915923  jessica.reyes@grimes.net      33615   1327730  2023-05-02   
12    5016915923  jessica.reyes@grimes.net      33615   6059361  2023-05-16   
724   5016915923  jessica.reyes@grimes.net      33615   7417275  2023-05-24   
840   5016915923  jessica.reyes@grimes.net      33615   1951089  2023-02-15   
1405  5016915923  jessica.reyes@grimes.net      33615   5615970  2023-04-19   

      order_week  order_year         item_id     category     subcategory  \
5             18        2023  BMB23158-88-9A        decor          plants   
12            20        2023  BDC65078-60-7M  consumables  paper products   
724           21        2023  ZBZ76765-83-2X  electronics         telecom   
840            7        2023  XBU40041-09-5M  consumables  paper products   
1405          16        2023  UCZ01380-66-1U  consumables  paper products   

      unit_price  unit_cost  unit_weight  qty  line_number  
5         172.84     154.44         2.79   78            0  
12        229.74     198.38        27.51  568            2  
724       222.03     155.57         1.33   26            2  
840         9.09       7.18         3.11   36            3  
1405       34.82      25.80         2.26   17            4

In [15]:
# Part 3 :: Sum "qty" for lead client
topclient_df["qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [16]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
subtotal_price = df["qty"] * df["unit_price"]
df["Price_Subtotal"] = subtotal_price
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  Price_Subtotal  
0     1096.80     762.71         7.50  105            1       115164.00  
1       24.95      15.09         1.49   21            0          523.95  
2       13.52       7.86         1.68   39            6          527.28  
3       36.42      24.85         1.23   29            3         1056.18  
4      195.10     108.17        46.43   20            1         3902.00

In [17]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

# First, add a column for "Total_Weight"
total_weight = df["qty"] * df["unit_weight"]
df["Total_Weight"] = total_weight
# Then, define a function for less than 50 pounds or equal or greater than 50
# Define a function
def total_shipping(shipping_weight):
    if shipping_weight < 50:
        return shipping_weight * 10
    return shipping_weight * 7

# "Apply" the function to the amount column
df['Total_Shipping'] = df['Total_Weight'].apply(total_shipping)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  Price_Subtotal  Total_Weight  \
0     762.71         7.50  105            1       115164.00        787.50   
1      15.09         1.49   21            0          523.95         31.29   
2       7.86         1.68   39            6          527.28         65.52   
3      24.85         1.23   29            3         1056.18         35.67   
4     108.17        46.43   20            1         3902.00        928.60   

   Total_Shipping  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [18]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
sales_tax_01 = .0925
total_price = ( df["Price_Subtotal"] + df["Total_Shipping"] )* (1 + sales_tax_01)
df["Total_Price"] = total_price
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  Price_Subtotal  Total_Weight  \
0         7.50  105            1       115164.00        787.50   
1         1.49   21            0          523.95         31.29   
2         1.68   39            6          527.28         65.52   
3         1.23   29            3         1056.18         35.67   
4        46.43   20            1         3902.00        928.60   

   Total_Shipping    Total_Price  
0         5512.50  131839.076250  
1          312.90     914.258625  
2          458.64    1077.117600  
3          356.70    1543.571400  
4         6500.20   11364.403500  

[5 rows x 22 columns]

In [19]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
total_line_cost = df["unit_cost"] * df["qty"] + df["Total_Shipping"]
df["Total_Line_Cost"] = total_line_cost
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  Price_Subtotal  Total_Weight  Total_Shipping    Total_Price  \
0            1       115164.00        787.50         5512.50  131839.076250   
1            0          523.95         31.29          312.90     914.258625   
2            6          527.28         65.52          458.64    1077.117600   
3            3         1056.18         35.67          356.70    1543.571400   
4            1         3902.00        928.60         6500.20   11364.403500   

   Total_Line_Cost  
0         85597.05  
1           629.79  
2           765.18  
3          1077.35  
4          8663.60  

[5 rows x 23 columns]

In [20]:
# Create a column for the profit of
# each line using line cost and line price
line_profit = df["Total_Price"] - df["Total_Line_Cost"]
df["Line_Profit"] = line_profit
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   Price_Subtotal  Total_Weight  Total_Shipping    Total_Price  \
0       115164.00        787.50         5512.50  131839.076250   
1          523.95         31.29          312.90     914.258625   
2          527.28         65.52          458.64    1077.117600   
3         1056.18         35.67          356.70    1543.571400   
4         3902.00        928.60         6500.20   11364.403500   

   Total_Line_Cost   Line_Profit  
0         85597.05  46242.026250  
1           629.79    284.468625  
2           765.18    311.937600  
3          1077.35    466.221400  
4          8663.60   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [21]:
# Check your work using the totals above
# Step 1: Input Order ID  ATTENTION CHECKER INPUT FUNCTION!!!
# Create a new df with the order_id(s)
# Verify the order_id is assigned to only 1 user (fault detection).

#Request first order input
order_1 = int(input("What is the first order number to calculate the Total Price?  "))
order_2 = int(input("What is the second order number to calculate the Total Price?  "))
order_3 = int(input("What is the third order number to calculate the Total Price?  "))
# INPUT >>>>>  order_1 = 2742071, order_2 = 2173913, order 3 = 6128929   <<<<<
# Create new dataframe and View Order 1 Data
order_1_df = df.loc[df['order_id'] == order_1]
order_1_df.head(3)



first   last                                       job  \
3446   Kimberly  Blair  Geographical information systems officer   
6287   Kimberly  Blair  Geographical information systems officer   
20623  Kimberly  Blair  Geographical information systems officer   

                    phone                           email  client_id  \
3446   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
6287   477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   
20623  477-680-9323x27378  blair.kimberly.6618@nelson.biz      78447   

       order_id  order_date  order_week  order_year  ... unit_cost  \
3446    2742071  2023-02-27           9        2023  ...      3.05   
6287    2742071  2023-02-27           9        2023  ...    407.68   
20623   2742071  2023-02-27           9        2023  ...     80.78   

      unit_weight  qty  line_number  Price_Subtotal  Total_Weight  \
3446        10.46   77            5          430.43        805.42   
6287         2.60  260            0       119652.00        676.00   
20623        2.27   12            3         1621.44         27.24   

       Total_Shipping    Total_Price  Total_Line_Cost   Line_Profit  
3446          5637.94    6629.694225          5872.79    756.904225  
6287          4732.00  135889.520000        110728.80  25160.720000  
20623          272.40    2069.020200          1241.76    827.260200  

[3 rows x 24 columns]

In [22]:
# Verify that this order number is only 1 purchaser (EXTRA fault detection)
order_1_df['client_id'].unique()

array([78447], dtype=int64)

In [23]:
# View Order 2 data
order_2_df = df.loc[df['order_id'] == order_2]
order_2_df.head(3)

first     last                 job                 phone  \
627    Darius  Stevens  Physicist, medical  +1-220-240-7019x2774   
14089  Darius  Stevens  Physicist, medical  +1-220-240-7019x2774   
14938  Darius  Stevens  Physicist, medical  +1-220-240-7019x2774   

                              email  client_id  order_id  order_date  \
627    stevensdarius@rich-leach.net      22122   2173913  2023-02-19   
14089  stevensdarius@rich-leach.net      22122   2173913  2023-02-19   
14938  stevensdarius@rich-leach.net      22122   2173913  2023-02-19   

       order_week  order_year  ... unit_cost unit_weight qty  line_number  \
627             7        2023  ...     96.23        2.66  70            4   
14089           7        2023  ...      2.09        1.04  36            2   
14938           7        2023  ...    160.50        4.08  23            6   

       Price_Subtotal  Total_Weight  Total_Shipping  Total_Price  \
627           7718.90        186.20         1303.40  9856.862750   
14089          118.08         37.44          374.40   538.034400   
14938         5835.33         93.84          656.88  7092.739425   

       Total_Line_Cost  Line_Profit  
627            8039.50  1817.362750  
14089           449.64    88.394400  
14938          4348.38  2744.359425  

[3 rows x 24 columns]

In [24]:
# Verify that this order number is only 1 purchaser (EXTRA fault detection)
order_2_df['client_id'].unique()

array([22122], dtype=int64)

In [25]:
# View Order 3 data
order_3_df = df.loc[df['order_id'] == order_3]
order_3_df.head(3)

first  last                       job                   phone  \
3422   Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
13672  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   
14517  Dorothy  Hall  Journalist, broadcasting  001-338-609-8495x03504   

                                        email  client_id  order_id  \
3422   dorothy.per.1343@mitchell-chambers.com      88257   6128929   
13672  dorothy.per.1343@mitchell-chambers.com      88257   6128929   
14517  dorothy.per.1343@mitchell-chambers.com      88257   6128929   

       order_date  order_week  order_year  ... unit_cost unit_weight   qty  \
3422   2023-03-12          10        2023  ...     11.75        7.61   671   
13672  2023-03-12          10        2023  ...     90.12        7.05    40   
14517  2023-03-12          10        2023  ...     98.35        6.28  1089   

       line_number  Price_Subtotal  Total_Weight  Total_Shipping  \
3422             3         8890.75       5106.31        35744.17   
13672            6         5459.60        282.00         1974.00   
14517            9       117187.29       6838.92        47872.44   

         Total_Price  Total_Line_Cost   Line_Profit  
3422    48763.650100         43628.42   5135.230100  
13672    8121.208000          5578.80   2542.408000  
14517  180327.755025        154975.59  25352.165025  

[3 rows x 24 columns]

In [26]:
# Verify that this order number is only 1 purchaser (EXTRA fault detection)
order_3_df['client_id'].unique()

array([88257], dtype=int64)

In [27]:
# Sum the Total_Price Column for Each Order Number Total Price
order_1_total_price = order_1_df["Total_Price"].sum()
order_2_total_price = order_2_df["Total_Price"].sum()
order_3_total_price = order_3_df["Total_Price"].sum()
# print(f"Order No. {order_1} Total Price is $ {order_1_total_price: .2f}")
# Multiline f-strings
print(
    f"Order No. {order_1} Total Price is $ {order_1_total_price: .2f}\n"
    + f"Order No. {order_2} Total Price is $ {order_2_total_price: .2f}\n"
    + f"Order No. {order_3} Total Price is $ {order_3_total_price: .2f}"
)



Order No. 2742071 Total Price is $  152811.90
Order No. 2173913 Total Price is $  162388.72
Order No. 6128929 Total Price is $  923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [28]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
# step 1 - test "top_five_clients" - Review:
numbers = [0,1,2,3,4]
for num in numbers:
    print(top_five_clients[num])

33615
66037
46820
38378
24741


In [29]:
# Step 2- Set up dataframes for each top 5 user.
#Note:   list "top_five_clients" previously defined
# Top User #1
client_1_df = df.loc[df['client_id'] == top_five_clients[0]]
client_1_df.head(3)

first   last                                           job       phone  \
5    Jessica  Reyes  Production designer, theatre/television/film  5016915923   
12   Jessica  Reyes  Production designer, theatre/television/film  5016915923   
724  Jessica  Reyes  Production designer, theatre/television/film  5016915923   

                        email  client_id  order_id  order_date  order_week  \
5    jessica.reyes@grimes.net      33615   1327730  2023-05-02          18   
12   jessica.reyes@grimes.net      33615   6059361  2023-05-16          20   
724  jessica.reyes@grimes.net      33615   7417275  2023-05-24          21   

     order_year  ... unit_cost unit_weight  qty  line_number  Price_Subtotal  \
5          2023  ...    154.44        2.79   78            0        13481.52   
12         2023  ...    198.38       27.51  568            2       130492.32   
724        2023  ...    155.57        1.33   26            2         5772.78   

     Total_Weight  Total_Shipping   Total_Price  Total_Line_Cost  Line_Profit  
5          217.62         1523.34   16392.80955         13569.66   2823.14955  
12       15625.68       109379.76  262060.24740        222059.60  40000.64740  
724         34.58          345.80    6684.54865          4390.62   2293.92865  

[3 rows x 24 columns]

In [30]:
# Step 3 - Dataframes for Remaining Users:
client_2_df = df.loc[df['client_id'] == top_five_clients[1]]
client_3_df = df.loc[df['client_id'] == top_five_clients[2]]
client_4_df = df.loc[df['client_id'] == top_five_clients[3]]
client_5_df = df.loc[df['client_id'] == top_five_clients[4]]
client_5_df.head(3)

first     last                     job         phone  \
416  Kendra  Garrett  Pension scheme manager  808.257.2029   
520  Kendra  Garrett  Pension scheme manager  808.257.2029   
671  Kendra  Garrett  Pension scheme manager  808.257.2029   

                               email  client_id  order_id  order_date  \
416  garrettkendra8479@jefferson.com      24741   3699423  2023-05-18   
520  garrettkendra8479@jefferson.com      24741   7980819  2023-05-25   
671  garrettkendra8479@jefferson.com      24741   9538464  2023-03-28   

     order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
416          20        2023  ...      8.17        0.96   65            5   
520          21        2023  ...     11.82        0.11  214            2   
671          13        2023  ...     10.21        2.77   30            1   

     Price_Subtotal  Total_Weight  Total_Shipping  Total_Price  \
416          596.05         62.40           436.8  1128.388625   
520         3513.88         23.54           235.4  4096.088400   
671          441.90         83.10           581.7  1118.283000   

     Total_Line_Cost  Line_Profit  
416           967.85   160.538625  
520          2764.88  1331.208400  
671           888.00   230.283000  

[3 rows x 24 columns]

In [31]:
# Step 4 & Final - # How much did each of the top 5 clients by quantity
# spend?  Sum each column for each dataframe and output.
print(
    f"The total spent by Top Five Qty Clients are \n"
    +f"{top_five_clients[0]}: $ {client_1_df['Total_Price'].sum():.2f}\n"
    +f"{top_five_clients[1]}: $ {client_2_df['Total_Price'].sum():.2f}\n"
    +f"{top_five_clients[2]}: $ {client_3_df['Total_Price'].sum():.2f}\n"
    +f"{top_five_clients[3]}: $ {client_4_df['Total_Price'].sum():.2f}\n"
    +f"{top_five_clients[4]}: $ {client_5_df['Total_Price'].sum():.2f}\n"
    )

The total spent by Top Five Qty Clients are 
33615: $ 8377144.70
66037: $ 10259514.80
46820: $ 9743794.32
38378: $ 12906550.88
24741: $ 82268891.98



In [32]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.


# Step 1: Create Lists for Each Header
column = "qty"
top_five_qty = [client_1_df[column].sum(),
                client_2_df[column].sum(),
                client_3_df[column].sum(),
                client_4_df[column].sum(),
                client_5_df[column].sum(),
                ]
print(top_five_qty)
#
column = "Total_Shipping"
top_five_shipping = [client_1_df[column].sum(),
                client_2_df[column].sum(),
                client_3_df[column].sum(),
                client_4_df[column].sum(),
                client_5_df[column].sum(),
                ]
print(top_five_shipping)
#
column = "Total_Price"
top_five_total_price = [client_1_df[column].sum(),
                client_2_df[column].sum(),
                client_3_df[column].sum(),
                client_4_df[column].sum(),
                client_5_df[column].sum(),
                ]
print(top_five_total_price)
#
column = "Total_Line_Cost"
top_five_total_line_cost = [client_1_df[column].sum(),
                client_2_df[column].sum(),
                client_3_df[column].sum(),
                client_4_df[column].sum(),
                client_5_df[column].sum(),
                ]
print(top_five_total_line_cost)
#
column = "Line_Profit"
top_five_line_profit = [client_1_df[column].sum(),
                client_2_df[column].sum(),
                client_3_df[column].sum(),
                client_4_df[column].sum(),
                client_5_df[column].sum(),
                ]
print(top_five_line_profit)

[64313, 43018, 75768, 73667, 239862]
[1828834.89, 1395151.8499999999, 1601448.8399999999, 3429455.4000000004, 5126448.37]
[8377144.697500001, 10259514.801275, 9743794.320474999, 12906550.880874999, 82268891.97985002]
[6175163.91, 7004482.98, 7007191.640000001, 9634720.979999999, 45688899.71]
[2201980.7875000006, 3255031.8212750005, 2736602.6804749994, 3271829.900875001, 36579992.26985001]


In [33]:
# Step 2:  Create a summary dataframe

top_5_summary_df_test = pd.DataFrame({
    "client_id": top_five_clients,
    "qty": top_five_qty,
    "Total_Shipping": top_five_shipping,
    "Total_Price": top_five_total_price,
    "Total_Cost": top_five_total_line_cost,
    "Line_Profit": top_five_line_profit
})
top_5_summary_df_test.head()

client_id     qty  Total_Shipping   Total_Price   Total_Cost   Line_Profit
0      33615   64313      1828834.89  8.377145e+06   6175163.91  2.201981e+06
1      66037   43018      1395151.85  1.025951e+07   7004482.98  3.255032e+06
2      46820   75768      1601448.84  9.743794e+06   7007191.64  2.736603e+06
3      38378   73667      3429455.40  1.290655e+07   9634720.98  3.271830e+06
4      24741  239862      5126448.37  8.226889e+07  45688899.71  3.657999e+07

In [61]:
# Format the data and rename the columns
# to names suitable for presentation.
# Step 1: Save the original dataframe and duplicate for formatted dataframe.
renamed_top_5_summary_df = top_5_summary_df_test.rename(columns=
{"client_id":"Client ID", "qty":"Units", "Total_Shipping": "Shipping", 
 "Total_Price": "Total Revenue", "Total_Cost": "Total Cost", 
 "Line_Profit": "Total Profit"})

# Currency should be in millions of dollars.
renamed_top_5_summary_df["Shipping"]=renamed_top_5_summary_df["Shipping"]/1000000
renamed_top_5_summary_df["Total Revenue"]=renamed_top_5_summary_df["Total Revenue"]/1000000
renamed_top_5_summary_df["Total Cost"]=renamed_top_5_summary_df["Total Cost"]/1000000
renamed_top_5_summary_df["Total Profit"]=renamed_top_5_summary_df["Total Profit"]/1000000

# renamed_top_5_summary_df['Shipping'] = renamed_top_5_summary_df.apply('Million ' + ['Shipping'], axis = 1)
renamed_top_5_summary_df.head()


Client ID   Units  Shipping  Total Revenue  Total Cost  Total Profit
0      33615   64313  1.828835       8.377145    6.175164      2.201981
1      66037   43018  1.395152      10.259515    7.004483      3.255032
2      46820   75768  1.601449       9.743794    7.007192      2.736603
3      38378   73667  3.429455      12.906551    9.634721      3.271830
4      24741  239862  5.126448      82.268892   45.688900     36.579992

In [62]:
#Add Currenty and "M" for Million
renamed_top_5_summary_df["Shipping"]=renamed_top_5_summary_df["Shipping"].apply(lambda x: "{}{:,.2f}{}".format('$', x, 'M'))
renamed_top_5_summary_df["Total Revenue"]=renamed_top_5_summary_df["Total Revenue"].apply(lambda x: "{}{:,.2f}{}".format('$', x, 'M'))
renamed_top_5_summary_df["Total Cost"]=renamed_top_5_summary_df["Total Cost"].apply(lambda x: "{}{:,.2f}{}".format('$', x, 'M'))
renamed_top_5_summary_df["Total Profit"]=renamed_top_5_summary_df["Total Profit"].apply(lambda x: "{}{:,.2f}{}".format('$', x, 'M'))
# Lambda formatting code from https://www.geeksforgeeks.org/formatting-integer-column-of-dataframe-in-pandas/

renamed_top_5_summary_df

Client ID   Units Shipping Total Revenue Total Cost Total Profit
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M

In [63]:
# Sort the updated data by "Total Profit" form highest to lowest
renamed_top_5_summary_df = renamed_top_5_summary_df.sort_values('Total Profit', ascending=False)
renamed_top_5_summary_df

Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M

# Summary
Client 24741 was significantly higher in revenue and profit than the next four client combined.  The shipping costs for Client 24741 were actually significantly lower per total revenue compared to the other top four clients.  Client 38378 had more than half the shipping costs of Client 24741 with a fraction of the sales.  CLient 38378 should try to reduce shipping costs to increase their revenue, or make sure they are passing these costs onto their customers if they are a pass through.